src/main/cpp/lang/xml.cpp

```diff
@@ -24,6 +23,42 @@ constexpr std::u8string_view comment_suffix = u8"-->";
 constexpr std::u8string_view illegal_comment_sequence = u8"--";
 constexpr std::u8string_view cdata_section_prefix = u8"<![CDATA[";
 constexpr std::u8string_view cdata_section_suffix = u8"]]>";
+constexpr std::u8string_view tag_suffix = u8">";
+
+struct End_Tag_Name_Stop {
+
+    bool operator()(std::u8string_view str) const
+    {
+        return !str.empty() && (str.starts_with(tag_suffix));
+    }
+};
+
+struct Start_Tag_Name_Stop {
+
+    bool operator()(std::u8string_view str) const
+    {
+        return !str.empty()
+            && (match_whitespace(str) || str.starts_with(u8"/>") || str.starts_with(tag_suffix));
+    }
+};
+
+struct Processing_Instr_Name_Stop {
+
+    bool operator()(std::u8string_view str) const
+    {
+        return !str.empty() && (match_whitespace(str) || str.starts_with(u8"?>"));
+    }
+};
+
+struct Attribute_Name_Stop {
+
+    bool operator()(std::u8string_view str) const
+    {
+        return !str.empty()
+            && (match_whitespace(str) || str.starts_with(u8'=') || str.starts_with(u8'>')
+                || str.starts_with(u8"/>"));
+    }
+};
```

Comment:
None of these need to be distinct types; can you please convert them to lambdas where they're used?